src/data/nat/count.lean

```diff
@@ -0,0 +1,559 @@
+`nat.subtype.order_iso_of_nat` should exist. See discussion
+[here](https://github.com/leanprover-community/mathlib/pull/9457#pullrequestreview-767221180).
+Future work should address how lemmas that use these should be written.
+
+-/
+
+lemma list.range_add (a : ℕ) :
+  ∀ b, list.range (a + b) = list.range a ++ (list.range b).map (λ x, a + x)
+| 0 := by rw [add_zero, list.range_zero, list.map_nil, list.append_nil]
+| (b + 1) := by rw [nat.add_succ, list.range_succ, list.range_add b, list.range_succ,
+  list.map_append, list.map_singleton, list.append_assoc]
+
+lemma multiset.range_add (a b : ℕ) :
+  multiset.range (a + b) = multiset.range a ∪ (multiset.range b).map (λ x, a + x) :=
+begin
+  rw [multiset.range, list.range_add, ←multiset.coe_add, ←multiset.coe_map],
+  refine multiset.add_eq_union_iff_disjoint.2 (λ x hxa hxb, _),
+  rw [multiset.mem_coe, list.mem_range] at hxa,
+  obtain ⟨c, _, rfl⟩ := multiset.mem_map.1 hxb,
+  exact (self_le_add_right _ _).not_lt hxa,
+end
+
+lemma finset.range_add (a b : ℕ) :
+  finset.range (a + b) = finset.range a ∪ (finset.range b).image (λ x, a + x) :=
+begin
+  rw [←finset.val_inj, finset.range, finset.union_val,
+    finset.image_val_of_inj_on ((add_right_injective a).inj_on _)],
+  exact multiset.range_add a b,
+end
+
+lemma galois_connection.lt_iff_lt {α β : Type*} [linear_order α] [linear_order β] {l : α → β}
+  {u : β → α} (gc : galois_connection l u) {a : α} {b : β} :
+  b < l a ↔ u b < a :=
+lt_iff_lt_of_le_iff_le (gc a b)
+
+open finset
+
+namespace nat
+variable (p : ℕ → Prop)
+
+section count
+variable [decidable_pred p]
+
+/-- Count the number of naturals `k < n` satisfying `p k`. -/
+def count (n : ℕ) : ℕ := (list.range n).countp p
+
+@[simp] lemma count_zero : count p 0 = 0 :=
+by rw [count, list.range_zero, list.countp]
+
+/-- A fintype instance for the set relevant to `nat.count`. Locally an instance in locale `count` -/
+def count_set.fintype (n : ℕ) : fintype {i // i < n ∧ p i} :=
+begin
+  apply fintype.of_finset ((finset.range n).filter p),
+  intro x,
+  rw [mem_filter, mem_range],
+  refl,
+end
+
+localized "attribute [instance] count_set.fintype" in count
+
+lemma count_eq_card_filter_range (n : ℕ) : count p n = ((range n).filter p).card :=
+by { rw [count, list.countp_eq_length_filter], refl,}
+
+/-- `count p n` can be expressed as the cardinality of `{k // k < n ∧ p k}`. -/
+lemma count_eq_card_fintype (n : ℕ) : count p n = fintype.card {k : ℕ // k < n ∧ p k} :=
+begin
+  rw [count_eq_card_filter_range, ←fintype.card_of_finset],
+  congr,
+  intro i,
+  rw [mem_filter, mem_range],
+  refl,
+end
+
+lemma count_monotone : monotone (count p) :=
+begin
+  intros a b h,
+  rw [count, list.countp_eq_length_filter, count, list.countp_eq_length_filter],
+  exact list.length_le_of_sublist (list.sublist.filter p (list.range_sublist.mpr h)),
+end
+
+@[simp] lemma count_succ (n : ℕ) : count p (n + 1) = count p n + (if p n then 1 else 0) :=
```

Comment:
I don't think this should be in the default simp-set.

Reply:
Ok, I removed it